### DISCUSSION TITLE
Scheduled posts from timeline yaml

### DISCUSSION THREAD

**Discussion by andrewpbray** [quarto-dev/quarto-cli#3584] [Q&A]
I have a blog that I'd like to publish on a schedule, with the publishing dates stored in a file that's separate from the posts. It seems like the scheduling question would be well-addressed by the filter described in #3114. What is the best approach for propagating the date information into the documents or into the render process?

I've put together a repo that explores a couple of different approaches to this. 

<https://github.com/andrewpbray/scheduled-posts-options>

They're described in the README and I've copied them here below. It strikes me that this is a version of a more general question: **what's the best way to render documents subject to document-specific metadata that is held in a separate file?**

The three approaches below reference a `_schedule.yml` that would look something like this:

```
schedule:
  - filename: "topic-1"
    date: "01/01/22"
  - filename: "topic-2"
    date: "01/02/22"
  - filename: "topic-3""
    date: "01/03/22"
```

#### 1. YML injection via .md file created in pre-render script

Write a pre-render script that parses `_schedule.yml` and creates a tiny `_publish-date.md` file next to each post that adds to the metadata via an include. This is quasi-implemented in Topic 2.

Downsides:
1.  feels kludgy
2.  generates a bunch of files, though this could be addressed by
     - adding files to .gitignore
     - keep files in dedicated include directory
     - create a post-render script to clean out that directory

#### 2. Reference _variables.yml in document YAML

Create a `_variables.yml` file that includes the publish dates for each post. That variable is then referenced in the yaml header of the post. This is implemented in Topic 3 (though apparently I've formatted the date field incorrectly).

A similar use case is discussed [here](https://github.com/quarto-dev/quarto-cli/discussions/2865).

Downsides:
1. Requires tailoring the yaml of each doc to reference the appropriate date.
2. Seems like a misuse of variables. These are meant to be shared across documents, right? Here they're document specific.

#### 3. Write a filter

Changing the meta data of a documents seems solidly within the functionality of a filter. The wrinkles that I see here:

1. The filter will need to reference information in `_schedule.yml`. Is there an example of a filter that does this that I could look at?
2. Right now, `_schedule.yml` links the publish date with the file / directory name of the post. I'm not sure how to reference that information in the filter. Would it make more sense to reference the `title` field in the yaml?

**Comment by srvanderplas** [comment]
This would also be very useful for e.g. course websites, where you don't want to overwhelm students at the beginning of the semester (so you only want content that is initially relevant) but you have the whole course prepped ahead of time. It would also make changing dates for an entire semester at once much easier.

**Comment by andrewpbray** [comment]
@srvanderplas I figured out a way to implement this using R scripts that are run before I run `quarto render`. The repo of course materials is [here](https://github.com/stat20/stat20) and the file that controls the publish dates is called [`_course-settings.yml`](https://github.com/stat20/stat20/blob/main/_course-settings.yml). I render the course website through github actions found [here](https://github.com/stat20/stat20/actions/workflows/publish-stat20.yml), which runs [this script](https://github.com/stat20/stat20/blob/main/assets/scripts/ignore-future-docs.r) before rendering.

The key idea of the script is to read the yml file and prepend `_` to any files with publish dates in the future. This has the effect of ignoring these files during the render process.

**Reply by mcanouil** [reply]
You could also add the files in. `project.render` using negation, see <https://quarto.org/docs/websites/index.html#render-targets>.
This means only one file to modify.

**Reply by andrewpbray** [reply]
I believe when you use that approach, the ignoring happens after quarto builds the document listings. That means that pages like the problem sets and labs have blank spots.
